### PR TITLE
Improve shadow delta handling

### PR DESCRIPTION
**Pull request by plskeggs**
nRF Cloud Multi Service Sample:
Enable PSM by default.
Increase connection_mgr stack size to prevent crashes with Wi-Fi.
Do not count server-side CoAP errors against send_failure_count, or else we disconnect from the cloud for no reason.
Clean up check_shadow function.

nRF Cloud Library:
Make logging and alert control section usable for CoAP in addition to MQTT.
Add error checking and handling to bad shadow delta values for control section.
Add nrf_cloud_shadow_delta_response_encode() for library users to accept or reject custom shadow deltas.
Improve error logging for CoAP so that server side errors get displayed.  An example is the JSON error returned when a customer exceeds their monthly quota for messages or location services.

Jira: IRIS-6800

### CODE REVIEW

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -394,6 +395,11 @@ Cellular samples (renamed from nRF9160 samples)
     * From using the :ref:`lte_lc_readme` library directly to using Zephyr's ``conn_mgr`` and the :kconfig:option:`CONFIG_LTE_CONNECTIVITY` Kconfig option.
     * The sample to remove redundant shadow updates for nRF Cloud.
     * Build instructions, board files, and DTC overlay file so that Wi-Fi scanning works for the nRF9161 DK and the nRF9160 DK.
+    * The default value of the Kconfig option :kconfig:option:`CONFIG_POWER_SAVING_MODE_ENABLE` to enabled.
+
+  * Fixed:
+
+    * Legitimate server side errors are not counted now as a reason to disconnect and reconnect to the cloud.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:402]:
```suggestion
    * Legitimate server side errors are not counted now as a reason to disconnect from and reconnect to the cloud.
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -755,19 +761,22 @@ Libraries for networking
...
     * :c:func:`nrf_cloud_fota_job_start` function that starts a FOTA update job.
+    * :c:func:`nrf_cloud_shadow_delta_response_encode()` to help accept or reject shadow delta desired settings.
 
   * Updated:
 
     * JSON manipulation moved from :file:`nrf_cloud_fota.c` to :file:`nrf_cloud_codec_internal.c`.
     * :c:func:`nrf_cloud_obj_location_request_create` to use the new function :c:func:`nrf_cloud_obj_location_request_payload_add`.
     * Retry handling for P-GPS data download errors to retry ``ECONNREFUSED`` errors.
     * By default, Wi-Fi location requests include only the MAC address and RSSI value.
+    * The shadow desired section for the config subsection is no longer deleted. Applications and samples should use the function :c:func:`nrf_cloud_shadow_delta_response_encode()` to prevent recurring deltas.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:772]:
```suggestion
    * The shadow desired section for the config subsection is no longer deleted.
      Applications and samples should use the function :c:func:`nrf_cloud_shadow_delta_response_encode()` to prevent recurring deltas.
```

File: include/net/nrf_cloud_codec.h [hunk 3]
```diff
@@ -645,6 +645,25 @@ int nrf_cloud_error_msg_decode(const char * const buf,
 			       const char * const msg_type,
 			       enum nrf_cloud_error * const err);
 
+/**
+ *  @brief Convert input shadow delta to output shadow reported or desired
```
Comment (peknis) [inline include/net/nrf_cloud_codec.h:649]:
```suggestion
 *  @brief Convert input shadow delta to output shadow reported or desired.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Comment by plskeggs** [comment]
@peknis - I made your changes.  Thanks for the review.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c [hunk 4]
```diff
@@ -1004,6 +794,335 @@ int json_send_to_cloud(cJSON *const request)
...
+}
+
+int nrf_cloud_device_control_update(const struct nrf_cloud_data *const in_data,
+				    struct nrf_cloud_data *out_data,
+				    enum nrf_cloud_ctrl_status *status)
+{
+	int err;
+	struct nrf_cloud_ctrl_data ctrl_data;
+
+	if ((in_data == NULL) || (out_data == NULL) || (status == NULL)) {
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c:1062]: i see the `-ESRCH` check below, so maybe `in_data->ptr == NULL` is just not considered an error?

File: samples/cellular/nrf_cloud_multi_service/Kconfig [hunk 5]
```diff
@@ -295,7 +295,7 @@ config COAP_SHADOW_CHECK_RATE_SECONDS
 
 config COAP_SHADOW_THREAD_STACK_SIZE
 	int "CoAP Shadow Thread Stack Size (bytes)"
-	default 2048
```
Comment (glarsennordic) [inline samples/cellular/nrf_cloud_multi_service/Kconfig:298]: commit log says enable PSM by default, but I don't see it anywhere, possibly left out by mistake?

Comment (glarsennordic) [reply samples/cellular/nrf_cloud_multi_service/Kconfig:298]: Also, is this the PR where you were going to remove the redundant PSM KConfig?

File: samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c [hunk 6]
```diff
@@ -34,32 +37,20 @@ static int check_shadow(void)
...
-			} else {
-				LOG_DBG("Delta acknowledged");
-			}
-		} else {
-			LOG_INF("Checking again in %d seconds",
-				CONFIG_COAP_SHADOW_CHECK_RATE_SECONDS);
-			err = -EAGAIN;
-		}
+	in_data.len = strlen(buf);
+	if (!in_data.len) {
```
Comment (glarsennordic) [inline samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c:47]: Maybe add a comment somewhere that `nrf_cloud_coap_shadow_get` returns an empty string when one isn't available?

Eventually it might be nice to handle this with callbacks. Offer an `nrf_cloud_coap_check_shadow` function that raises the same callbacks that MQTT raises if a shadow is found. Then we can avoid stuff like checking for an empty string, plus it would mean our event pathway would be more similar between MQTT and COAP

Comment (glarsennordic) [reply samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c:47]: I think we already have tickets covering that idea, of course, just pointing it out here for sake of discussion

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 7]
```diff
@@ -395,6 +396,10 @@ Cellular samples (renamed from nRF9160 samples)
     * The sample to remove redundant shadow updates for nRF Cloud.
     * Build instructions, board files, and DTC overlay file so that Wi-Fi scanning works for the nRF9161 DK and the nRF9160 DK.
 
+  * Fixed:
+
+    * Legitimate server side errors are not counted now as a reason to disconnect from and reconnect to the cloud.
```
Comment (glarsennordic) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:401]: Can you elaborate on this? Maybe I missed it, but I didn't see any logic changes regarding disconnect due to server side errors. But if there is a server-side error ongoing, to me it seems valid to disconnect and try again later -- unless you are talking specifically about non-fatal server-side errors that don't require a reconnect for the link with nRFCloud to continue functioning?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 8]
```diff
@@ -480,65 +362,69 @@ static int handle_pin_complete(const struct nct_evt *nct_evt)
...
 	const enum nfsm_state current_state = nfsm_get_current_state();
-
-	LOG_DBG("CC RX on topic %s: %s",
+	struct nct_cc_data msg = {
+		.opcode = NCT_CC_OPCODE_UPDATE_REQ,
+		.message_id = NCT_MSG_ID_STATE_REPORT
+	};
+	struct nrf_cloud_topic *t = &nct_evt->param.cc->topic;
+	int dlen = strlen(NRF_CLOUD_JSON_KEY_DELTA);
+	bool is_delta = (strncmp(&((const char *)t->ptr)[t->len - dlen],
```
[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:377]: I'm resolving this since it was kinda eclipsed by my other comments, and I think Justin has also commented on this

File: include/net/nrf_cloud_codec.h [hunk 9]
```diff
@@ -645,6 +645,25 @@ int nrf_cloud_error_msg_decode(const char * const buf,
...
+ *  When rejecting, the input_obj shadow_fragment should contain the previously
+ *  valid values for any invalid values received in the desired section.
+ *
+ *  @param[in]  input_obj  Shadow fragment to encode for sending.
+ *  @param[in]  accept     Flag to indicate whether to accept (place in reported section) or reject
+ *                         (place in desired section).
+ *  @param[out] output     Pointer to and length of a buffer containing the JSON-formatted
+ *                         text to send.
+ */
+int nrf_cloud_shadow_delta_response_encode(cJSON *input_obj,
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply include/net/nrf_cloud_codec.h:663]: That's where I originally needed it (in _internal) but decided to make it public.  I'll move it.

[the remark this answers is not in the document]
Comment (plskeggs) [reply include/net/nrf_cloud_codec.h:663]: Regarding nrf_cloud_obj, I'll try.

For an example, I originally wanted to use it from ATv2, but as I dug into it the work looked more complicated than I wanted to do at this time.

I suppose I could add a custom shadow handler in nrf_cloud_multi_service that would use this.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c [hunk 10]
```diff
@@ -1004,6 +794,335 @@ int json_send_to_cloud(cJSON *const request)
...
+	 * {"version":2413,"timestamp":1695404679,
+	 * "state":{"config":{"activeMode":false}},
+	 * "metadata":{"config":{"activeMode":{"timestamp":1695404646}}}}
+	 */
+	delta_obj = cJSON_DetachItemFromObject(input_obj, NRF_CLOUD_JSON_KEY_STATE);
+	cJSON_Delete(input_obj);
+
+	/* If this is not a delta update, no response data is required */
+	if (delta_obj == NULL) {
+		LOG_DBG("Not a delta");
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c:826]: The function name tells you where it came from (if that's enabled in the logging settings), but sure, I can make it more specific.

[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c:826]: Oh my bad, I thought we only had thread name logging

Still more specific is probably a good idea for those whoops moments where it wasn't enabled 😅

File: samples/cellular/nrf_cloud_multi_service/Kconfig (hunk 5 above)
Comment (glarsennordic) [reply samples/cellular/nrf_cloud_multi_service/Kconfig:298]: Aha, I see you've done this in https://github.com/nrfconnect/sdk-nrf/pull/12492 instead -- so then I think that mention just needs to be removed from the commit log

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c [hunk 11]
```diff
@@ -490,6 +499,13 @@ static int client_transfer(enum coap_method method,
 		(void)k_sem_take(&cb_sem, K_FOREVER); /* Wait for coap_client to exhaust retries */
 	}
 
+	if (!reliable && !err && (user_cb.result_code >= COAP_RESPONSE_CODE_BAD_REQUEST)) {
+		/* NON transfers usually do not use a callback,
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:503]: It is a CoAP-specific term; it is an abbreviation for non-confirmable transfers.  With a NON transfer, the sender does not expect an ACK packet back.  With a CON (confirmable) transfer, it does expect one back, and if it does not get one, retries the transfer with an exponentially increasing delay until the transfer timeout is exceeded (~30 seconds).

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 7 above)
Comment (plskeggs) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:401]: I'm talking about errors like getting a 4.04 error on a CoAP transfer indicating the resource requested could not be found.  There is no valid reason to disconnect from the cloud and reconnect, which wastes power and data.

The change is in message_queue.c line 134.

Currently MQTT transfers always return negative error codes or zero, so this changelog entry does not apply to that.

File: samples/cellular/nrf_cloud_multi_service/Kconfig (hunk 5 above)
Comment (plskeggs) [reply samples/cellular/nrf_cloud_multi_service/Kconfig:298]: Good catch -- I removed the change I had made to the Kconfig to default to y (which did nothing, as we discussed) but forgot to remove the comment.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 7 above)
Comment (plskeggs) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:401]: I'll improve the comment.

File: samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c (hunk 6 above)
Comment (plskeggs) [reply samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c:47]: There is a comment in nrf_cloud_coap.h already.

Go ahead and add to the existing ticket.  I'm not convinced that merging the CoAP and MQTT code together long term is a good idea, in that the CoAP code is fairly simple compared to the MQTT.  Merging them will make both more complex.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 7 above)
Comment (glarsennordic) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:401]: Ooh, gotcha. That delta totally flew under my radar. Thanks for the clarification

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c [hunk 12]
```diff
@@ -490,6 +499,13 @@ static int client_transfer(enum coap_method method,
...
+		 * so make sure a bad result is not ignored.
+		 */
+		err = user_cb.result_code;
+	}
+
 	k_sem_give(&serial_sem);
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:509]: All the functions in nrf_cloud_coap.c call this function, and there is no other control to prevent a multi-threaded application from calling multiple such functions simultaneously.  We only use a single coap_client instance, and each instance only supports processing one transfer at a time, so we need the semaphore to ensure this.

File: samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c (hunk 6 above)
Comment (glarsennordic) [reply samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c:47]: Alright, I'll add it

Maybe you are right that merging them completely isn't a good idea, but I do think having both use event callbacks is worth considering. It could simplify applications that support both without too much complexity cost library-side IMO

Comment (glarsennordic) [reply samples/cellular/nrf_cloud_multi_service/src/shadow_support_coap.c:47]: Looks like the relevant ticket is IRIS-6790

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c:1062]: Yes, it is not an error to not have a control section in the shadow.  This function is called to detect and process a control section if present.

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:509]: Oh, duh. You know, like, the most basic use-case for semaphores. My bad

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 13]
```diff
@@ -480,65 +362,69 @@ static int handle_pin_complete(const struct nct_evt *nct_evt)
...
+				 NRF_CLOUD_JSON_KEY_DELTA, dlen) == 0);
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:378]: I'm removing it.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 14]
```diff
@@ -480,65 +362,69 @@ static int handle_pin_complete(const struct nct_evt *nct_evt)
...
+	enum nrf_cloud_ctrl_status status;
+
+	LOG_DBG("CC %sRX on topic %s: %s",
+		is_delta ? "DELTA " : "",
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:382]: I'm removing it.

[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:382]: Again, removing it.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:377]: And why is there an nct_cc_rx_opcode_map and not also an ncr_cc_tx_opcode_map, since this map is used for both RX and TX lists?

I don't see that any part of sdk-nrf uses the RX opcode, but the TX opcode is used.

I'll add a tx map and fix this up.

[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:377]: Nevermind.  I'll create ticket to follow up on that.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 15]
```diff
@@ -480,65 +362,69 @@ static int handle_pin_complete(const struct nct_evt *nct_evt)
...
+			LOG_DBG("Confirming shadow: %s",
+				(const char *)msg.data.ptr);
+			err = nct_cc_send(&msg);
+			nrf_cloud_free((void *)msg.data.ptr);
+			if (err) {
+				LOG_ERR("nct_cc_send failed %d", err);
+			}
+		}
+	} else {
+		LOG_ERR("Error: %d, msg.data.ptr: %p", err, msg.data.ptr);
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:399]: not anymore; removing

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:377]: yeah i think the rx/tx topic op codes and lists could be re-worked, so probably better to do in a separate ticket.
it is very hard to follow.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 16]
```diff
@@ -547,6 +433,7 @@ static int cc_rx_data_handler(const struct nct_evt *nct_evt)
 	case STATE_DC_CONNECTING:
 	case STATE_DC_CONNECTED:
 		if (new_state == STATE_UA_PIN_WAIT) {
+			/* Device was just removed from its nRF Cloud account. */
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:436]: It was fully connected, in the `STATE_DC_CONNECTED` state, and is now back to the `STATE_UA_PIN_WAIT` state.  So I think that only happens when a device is removed from the user's account.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c [hunk 17]
```diff
@@ -1004,6 +794,335 @@ int json_send_to_cloud(cJSON *const request)
...
+						 accept ? NRF_CLOUD_JSON_KEY_REP :
+							  NRF_CLOUD_JSON_KEY_DES);
+
+	/* When accepting the change, add the delta to reported:
+	 *   accept: {"state":{"reported":{"config":{"activeMode":false}}}}
+	 * When rejecting, either set the desired to null or the last valid value,
+	 * such as if sent a negative value where only positive values are valid:
+	 *   reject: {"state":{"desired":{"config":{"locationTimeout":400}}}}
+	 */
+	 output_obj = cJSONUtils_MergePatch(response_obj, delta_obj);
```
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c:849]: Rewrote it.  It did work with merge patch, but it's not needed.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:377]: https://nordicsemi.atlassian.net/browse/IRIS-7258

File: include/net/nrf_cloud_codec.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (plskeggs) [reply include/net/nrf_cloud_codec.h:663]: It doesn't really look right in nrf_cloud_codec.c -- there are no other similar functions there, but there are in nrf_cloud_codec_internal.c.  Anyway, I moved it just now.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c]:
```
When rejecting, either set the desired to null or the last valid value,
such as if sent a negative value where only positive values are valid:
```
the code is adding the delta_obj to the desired when rejecting?

**Review by peknis — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c]: Yes.  Perhaps "delta_obj" is a misleading name.

If the application accepts fully the shadow delta, then delta_obj will be the same as the contents of state.desired.

If the application rejects fully the shadow delta, then delta_obj needs to be either null or a corrected copy of state.desired.

[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec_internal.c]: In this case, I did not implement a way to set it to null, so I should probably change the comment.

File: include/net/nrf_cloud_codec.h [hunk 18]
```diff
@@ -645,6 +645,43 @@ int nrf_cloud_error_msg_decode(const char * const buf,
...
+ *  If rejecting, the delta should be modified with the correct data and passed
+ *  into this function as the input_obj parameter, with the accept flag set to false.
+ *  Example input_obj:
+ *      "myData":{"myValue":3}
+ *
+ *  A value can be removed from the shadow by setting it to null.
+ *  Example input_obj:
+ *      "myData":{"myValue":null}
+ *  The output parameter can then be sent to nRF Cloud using nrf_cloud_coap_shadow_state_update()
+ *  when using CoAP.
```
Comment (jayteemo) [inline include/net/nrf_cloud_codec.h:673]:
```suggestion
 *  when using CoAP or nrf_cloud_send() when using MQTT.
```

Comment (plskeggs) [reply include/net/nrf_cloud_codec.h:673]: Thanks @jayteemo -- I hadn't gotten to this yet.

**Review by jayteemo — APPROVED** [review]

**Review by glarsennordic — APPROVED** [review]
